### ISSUE TITLE
Adding new VMDK in existing folder

### ISSUE THREAD

**Issue by donwlewis** [chef-partners/knife-vsphere#213]
I would like to be able to add a new VMDK in the same datastore that the VM resides in.  When I attempt to do this, I get the following error:

```
knife vsphere vm vmdk add t1-cmhf-pkgs-l001 5 -f Test --target-lun CHS-datastore01 -V
INFO: Using configuration from C:/Users/lewisd01/Code/prodtech/.chef/knife.rb
Enter your password:
Looking for all datastores that match /CHS-datastore01/
Datastore: CHS-datastore01 (751.90 GB(807351091200) / 2.00 TB)
ERROR: RbVmomi::Fault: FileAlreadyExists: Cannot complete the operation because the file or folder [CHS-datastore01] t1-cmhf-pkgs-l001 already exists
```

Is it possible to create a new disk in the same datastore as an existing disk?

**Comment by swalberg** [comment]
Hi Don, what you're asking for is something we should be able to do already. In [this line](https://github.com/chef-partners/knife-vsphere/blob/master/lib/chef/knife/vsphere_vm_vmdk_add.rb) we only create the directory if we think it doesn't exist. I think you're getting the exception on line 63. You could comment it out if you wanted to see for sure.

**Comment by donwlewis** [comment]
So commenting out the line seems to fix the issue.  What am I doing wrong then?

**Comment by swalberg** [comment]
Hi, thanks for doing that test. I don't think you're doing anything wrong. I think that the code is not checking for the existence of the file correctly.

I will do some testing next week to see if I can narrow down the problem. I don't know if it's specific to LUN based stores or not, so I will try to replicate it myself.

**Comment by donwlewis** [comment]
Awesome.  Thanks.  Let me know how I can help.  I am new to ruby so I will try the best I can.  Looking at the code, I wasn't sure if there was an attribute under the vmdk_datastore that would have the list of folders that are in said datastore, or if there would need to be a function created to get a list of all the folders in that datastore, and then see if that name existed in the array.

**Comment by cpressland** [comment]
I've just run into exactly the same thing, any news on if this'll get fixed in the master branch? I'm happy to provide any testing etc also.

**Comment by swalberg** [comment]
Hi, I intend to look at this soon. I've been traveling the past couple of week and am running behind on things :(

**Comment by cpressland** [comment]
Thats for the update, really appreciate it.

**Comment by swalberg** [comment]
I've been playing around with this... I'm not using a LUN but `#exists?` behaves as I'd expect...

```
(byebug) d.exists? "blah"
false
(byebug) d.exists? "seantest"
true
```

![vcenter 2015-12-24 08-43-48](https://cloud.githubusercontent.com/assets/221368/11995214/d1ab3664-aa1a-11e5-8c46-cb2f7285440e.jpg)

If you browse the datastore do you see the datastore as I do? Can you describe how the datastores are laid out? Are they part of a cluster?

**Comment by donwlewis** [comment]
Yes, I believe I have only tested VMs that were part of a datastore
cluster.  I will try and test one that is not and if the results are the
same.

**Comment by donwlewis** [comment]
So I tried this with a system that was not in a datastore cluster, and the functionality seems normal, so the issue must only be with systems that are part of a datastore cluster.

**Comment by swalberg** [comment]
Thanks for checking into that. I'll see if I can detect that, and failing that, we'll need some kind of override.

**Comment by cpressland** [comment]
Just re-tested this on a vSphere environment that didn't have a Datastore Cluster and it seemed to work fine. Our production system is in a Datastore Cluster but I'm happy to test anything you'll need to figure this one out.
<img width="1412" alt="screen shot 2016-01-07 at 11 48 40" src="https://cloud.githubusercontent.com/assets/6557665/12170014/3ecf82ac-b535-11e5-87af-a887409964c0.png">

**Comment by swalberg** [comment]
I'm not a VMWare administrator, so we're getting to my limit of understanding on storage here. Please set me straight if I go astray...

My system has a datastore cluster but the disks are NFS and this command works for me. I can run the code in byebug and `exists?` works against `vm.datastore` (the datastore the VM is on) and the other member in the datastore cluster (by passing `--target-lun` and specifying the other member of the cluster). We are running storage DRS if that matters.

So a couple of thoughts:

* Do you need to use `--target-lun`? The non LUN code path puts the disk on the same datastore as the VM, and also skips that check.
* Does `--target-lun` work if you specify the name of the datastore containing the existing disk?

This behaviour was added in c5768dd and unfortunately the author can't be found on GitHub.

If neither of those works, or aren't feasible, I think I'll make it catch the fault and continue rather than adding a command line option or change the behaviour.

**Comment by swalberg** [comment]
Can you guys have a look at #219 and let me know if you think that's a reasonable workaround for now?

**Comment by cpressland** [comment]
Just Finished Testing, looks like it works fine now. Awesome work, really appreciate it.

Before:

    cpressland@cpressland-mbp ~/Documents/Projects/glp-cb-nginx-loadbalancer/.chef (master●)$ knife vsphere vm vmdk add cptest01 10GB --vmdk-type thick --target-lun Datastore04
    
    Looking for all datastores that match /Datastore04/
    Datastore: Datastore04 (4.28 TB(4700952854528) / 5.00 TB)
    ERROR: RbVmomi::Fault: FileAlreadyExists: Cannot complete the operation because the file or folder [Datastore04] cptest01 already exists

After:

    cpressland@cpressland-mbp ~/Documents/Projects/glp-cb-nginx-loadbalancer/.chef (master●)$ knife vsphere vm vmdk add cptest01 10GB --vmdk-type thick --target-lun Datastore04
    
    Looking for all datastores that match /Datastore04/
    Datastore: Datastore04 (4.28 TB(4700952854528) / 5.00 TB)
    WARNING: Ignoring a fault when trying to create [Datastore04] cptest01. This may be related to issue #213.
    Choosing: Datastore04
    Next vmdk name is => [Datastore04] cptest01/cptest01_2.vmdk
    Creating VMDK
    Capacity: 10GB GB
    Disk: [Datastore04] cptest01/cptest01_2.vmdk
    Attaching VMDK to cptest01
    using SCSI controller 0
    using SCSI ID 1
    cpressland@cpressland-mbp ~/Documents/Projects/glp-cb-nginx-loadbalancer/.chef (master●)$ knife vsphere vm vmdk add cptest01 10GB --vmdk-type thick --target-lun Datastore04
    
    Looking for all datastores that match /Datastore04/
    Datastore: Datastore04 (4.25 TB(4672895057920) / 5.00 TB)
    WARNING: Ignoring a fault when trying to create [Datastore04] cptest01. This may be related to issue #213.
    Choosing: Datastore04
    Next vmdk name is => [Datastore04] cptest01/cptest01_3.vmdk
    Creating VMDK
    Capacity: 10GB GB
    Disk: [Datastore04] cptest01/cptest01_3.vmdk
    Attaching VMDK to cptest01
    using SCSI controller 0
    using SCSI ID 2
    cpressland@cpressland-mbp ~/Documents/Projects/glp-cb-nginx-loadbalancer/.chef (master●)$

**Comment by swalberg** [comment]
Thanks, Chris. I'll merge it and release a new version of the gem later today.

**Comment by swalberg** [comment]
Released version 1.2.7 to rubygems. `gem install knife-vsphere` should get it for you.
